src/cm/communication_manager.rs

```diff
@@ -98,13 +360,277 @@ impl CommunicationManager {
         Ok(())
     }
 
-    pub fn listen(&mut self, backlog: i32) -> Result<(), String> {
-        let ret = unsafe { rdma_listen(self.cm_id.as_mut(), backlog) };
+    pub fn resolve_addr(
+        &self, src_addr: Option<SocketAddr>, dst_addr: Option<SocketAddr>, timeout_ms: i32,
+    ) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+
+        let ret = unsafe {
+            rdma_resolve_addr(
+                cm_id.as_mut(),
+                match src_addr {
+                    Some(addr) => OsSocketAddr::from(addr).as_mut_ptr(),
+                    None => null_mut(),
+                },
+                match dst_addr {
+                    Some(addr) => OsSocketAddr::from(addr).as_mut_ptr(),
+                    None => null_mut(),
+                },
+                timeout_ms,
+            )
+        };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve address {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn resolve_route(&self, timeout_ms: i32) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+        let ret = unsafe { rdma_resolve_route(cm_id.as_mut(), timeout_ms) };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve route {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn listen(&self, backlog: i32) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+        let ret = unsafe { rdma_listen(cm_id.as_mut(), backlog) };
 
         if ret < 0 {
             return Err(format!("Failed to listen {ret}"));
         }
 
         Ok(())
     }
+
+    pub fn get_device_context(&self) -> Result<ManuallyDrop<DeviceContext>, String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+
+        unsafe {
+            Ok(ManuallyDrop::new(DeviceContext {
+                context: (*cm_id.as_ptr()).verbs,
+            }))
+        }
+    }
+
+    pub fn connect(&self, conn_param: &mut ConnectionParameter) -> Result<(), String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+        let ret = unsafe { rdma_connect(cm_id.as_ptr(), &mut conn_param.0) };
+
+        if ret < 0 {
+            return Err(format!("Failed to connect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn disconnect(&self) -> Result<(), String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+        let ret = unsafe { rdma_disconnect(cm_id.as_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to disconnect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept(&self, conn_param: Option<&mut ConnectionParameter>) -> Result<(), String> {
```

Comment:
Is `conn_param` a optional paramter in `rdma_accept`?

Reply:
Yes, see https://github.com/linux-rdma/rdma-core/blob/d56c01d43a9c278b5cdd801b55108cfc7653421b/librdmacm/examples/cmtime.c#L227 and https://github.com/RDMA-Rust/sideway/blob/9fc51dd327e19714cdda4f0056c0aaf374c718a4/examples/cmtime.rs#L370 it's runnable.

Reply:
But that requires us bind a `qp` with `cm_id`